In [1176]:
#importing the required libraries
import pandas as pd
import matplotlib.pyplot as plt
import re

In [1177]:
#importing and reading through the data
df = pd.read_csv('movies_data.csv')
df.head(10)

MOVIES         YEAR  \
0                        Blood Red Sky       (2021)   
1  Masters of the Universe: Revelation     (2021– )   
2                     The Walking Dead  (2010–2022)   
3                       Rick and Morty     (2013– )   
4                      Army of Thieves       (2021)   
5                          Outer Banks     (2020– )   
6      The Last Letter from Your Lover       (2021)   
7                               Dexter  (2006–2013)   
8                    Never Have I Ever     (2020– )   
9                         Virgin River     (2019– )   

                                        GENRE  RATING  \
0      \nAction, Horror, Thriller                 6.1   
1  \nAnimation, Action, Adventure                 5.0   
2       \nDrama, Horror, Thriller                 8.2   
3  \nAnimation, Adventure, Comedy                 9.2   
4         \nAction, Crime, Horror                 NaN   
5          \nAction, Crime, Drama                 7.6   
6                \nDrama, Romance                 6.8   
7         \nCrime, Drama, Mystery                 8.6   
8                        \nComedy                 7.9   
9                \nDrama, Romance                 7.4   

                                            ONE-LINE  \
0  \nA woman with a mysterious illness is forced ...   
1  \nThe war for Eternia begins again in what may...   
2  \nSheriff Deputy Rick Grimes wakes up from a c...   
3  \nAn animated series that follows the exploits...   
4  \nA prequel, set before the events of Army of ...   
5  \nA group of teenagers from the wrong side of ...   
6  \nA pair of interwoven stories set in the past...   
7  \nBy day, mild-mannered Dexter is a blood-spat...   
8  \nThe complicated life of a modern-day first g...   
9  \nSeeking a fresh start, nurse practitioner Me...   

                                               STARS    VOTES  RunTime Gross  
0  \n    Director:\nPeter Thorwarth\n| \n    Star...   21,062    121.0   NaN  
1  \n            \n    Stars:\nChris Wood, \nSara...   17,870     25.0   NaN  
2  \n            \n    Stars:\nAndrew Lincoln, \n...  885,805     44.0   NaN  
3  \n            \n    Stars:\nJustin Roiland, \n...  414,849     23.0   NaN  
4  \n    Director:\nMatthias Schweighöfer\n| \n  ...      NaN      NaN   NaN  
5  \n            \n    Stars:\nChase Stokes, \nMa...   25,858     50.0   NaN  
6  \n    Director:\nAugustine Frizzell\n| \n    S...    5,283    110.0   NaN  
7  \n            \n    Stars:\nMichael C. Hall, \...  665,387     53.0   NaN  
8  \n            \n    Stars:\nMaitreyi Ramakrish...   34,530     30.0   NaN  
9  \n            \n    Stars:\nAlexandra Breckenr...   27,279     44.0   NaN

## Making a Copy of the Dataset

In [1178]:
#making a copy of my dataframe
df_movies = df.copy()

In [1179]:
#getting the information about the data
df_movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9999 entries, 0 to 9998
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   MOVIES    9999 non-null   object 
 1   YEAR      9355 non-null   object 
 2   GENRE     9919 non-null   object 
 3   RATING    8179 non-null   float64
 4   ONE-LINE  9999 non-null   object 
 5   STARS     9999 non-null   object 
 6   VOTES     8179 non-null   object 
 7   RunTime   7041 non-null   float64
 8   Gross     460 non-null    object 
dtypes: float64(2), object(7)
memory usage: 703.2+ KB


In [1180]:
#decription of the data
df_movies.describe()

RATING      RunTime
count  8179.000000  7041.000000
mean      6.921176    68.688539
std       1.220232    47.258056
min       1.100000     1.000000
25%       6.200000    36.000000
50%       7.100000    60.000000
75%       7.800000    95.000000
max       9.900000   853.000000

In [1181]:
#columns with null values
df_movies.isnull().any()

MOVIES      False
YEAR         True
GENRE        True
RATING       True
ONE-LINE    False
STARS       False
VOTES        True
RunTime      True
Gross        True
dtype: bool

In [1182]:
#to check out for duplicates
df_movies.columns.duplicated()

array([False, False, False, False, False, False, False, False, False])

## Data Wrangling
>  This process involves tidying, cleaning, and ensuring the data is of good quality to generate insights.

In [1183]:
#information about the data
df_movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9999 entries, 0 to 9998
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   MOVIES    9999 non-null   object 
 1   YEAR      9355 non-null   object 
 2   GENRE     9919 non-null   object 
 3   RATING    8179 non-null   float64
 4   ONE-LINE  9999 non-null   object 
 5   STARS     9999 non-null   object 
 6   VOTES     8179 non-null   object 
 7   RunTime   7041 non-null   float64
 8   Gross     460 non-null    object 
dtypes: float64(2), object(7)
memory usage: 703.2+ KB


## Issue 1
#### Renaming columns names to achieve uniformity

In [1184]:
#renaming columns
df_movies = df_movies.rename(columns = {'MOVIES':'movies', 'YEAR':'year','GENRE':'genre',
                                        'RATING':'rating', 'ONE-LINE':'one_line','STARS':'stars','VOTES':'votes',
                                        'RunTime':'run_time', 'Gross':'gross'})
df_movies.sample()

movies    year                       genre  rating  \
998  Spinning Out  (2020)  \nDrama, Sport                 7.6   

                                              one_line  \
998  \nA figure skating Olympic hopeful's struggle ...   

                                                 stars   votes  run_time gross  
998  \n            \n    Stars:\nKaya Scodelario, \...  11,776      48.0   NaN

## Issue 2
#### Genre : Removing the extra white spaces, extracting the `genre` from the string using regular expression

In [1138]:
#converting the genre to str type for cleanliness
df_movies['genre'] = df_movies['genre'].astype(str)

In [1139]:
#removing the genre regex and newlines 
df_movies['genre']= df_movies['genre'].apply(lambda x: ",".join(x.strip().replace(" ","").split('\n')))
df_movies['genre']

0           Action,Horror,Thriller
1       Animation,Action,Adventure
2            Drama,Horror,Thriller
3       Animation,Adventure,Comedy
4              Action,Crime,Horror
                   ...            
9994       Adventure,Drama,Fantasy
9995    Animation,Action,Adventure
9996             Documentary,Sport
9997       Adventure,Drama,Fantasy
9998       Adventure,Drama,Fantasy
Name: genre, Length: 9999, dtype: object

## Issue 3
#### One_line: Stripping the column to extract the `one_line` column needed information

In [1140]:
#removing the extra white spaces in the one_line column
df_movies['one_line'] = df_movies['one_line'].apply(lambda x: x.strip())

## Issue 4
#### Director: Creating a new column called `director` in order to extract the `directors` details from the `stars` column details  

In [1141]:
#extracting the directors details using regex
df_movies['director'] = df_movies['stars'].apply(lambda x: re.sub("\s\s+" , " ", x.strip().split('|')[0].strip()))

In [1142]:
#function to confirm if there are directors in all the stars rows 
def director_search(x):
    if 'Director' in x: 
        director = x.split('\n')[1]
        return director
    else:
        return 'None'

In [1143]:
#applying the function to the directors column
df_movies['director'] = df_movies['director'].apply(director_search)

In [1144]:
#confirming we have only the directors name
df_movies['director'].value_counts()

None                                        3646
Rajiv Chilaka, Krishna Mohan Chintapatla      49
Adam Heydt                                    28
Baran bo Odar                                 23
Oliver Driver                                 21
                                            ... 
Anthony Stacchi                                1
Jing Wong, Wai-Lun Lam                         1
David Marconi                                  1
Shantrelle P. Lewis                            1
Ian Daniel, Elliot Page                        1
Name: director, Length: 3709, dtype: int64

## Issue 5
#### Stars : Getting the exact `stars` int the different movie set

In [1145]:
#function to get the stars info in the columns as some rows have stars while some dont
def stars_search(x):
    x = x.strip()
    if 'Director' in x: 
        x = x.split('|')
        if len(x) > 1:
            star = x[1]
        else:
            return 'None'          
    else:
        star = x
    star = star.strip()
    star = star.split('\n')[1:]
    star = ",".join(star)
    star = re.sub("\s\s+" , " ", star)
    star = star.split(',')
    star = list(filter(str.strip, star))
    star = ",".join(star)
    return star

In [1146]:
#creating a new colummn for movie stars
df_movies['movies_star'] = df_movies['stars'].apply(stars_search)

In [1147]:
#asserting and confirming this has been cleaned 
df_movies1 = df_movies.loc[df_movies['movies_star']== 'None']
df_movies1.head(2)

movies    year                      genre  \
1367  El caso Wanninkhof-Carabantes  (2021)  Documentary,Crime,History   
1469                      Pray Away  (2021)                Documentary   

      rating                                           one_line  \
1367     6.5  In 1999, teen Rocío Wanninkhof is murdered. He...   
1469     6.2  Former leaders of the "pray the gay away" move...   

                                     stars  votes  run_time gross  \
1367        \n    Director:\nTània Balló\n  1,357      88.0   NaN   
1469  \n    Director:\nKristine Stolakis\n    159     101.0   NaN   

               director movies_star  
1367        Tània Balló        None  
1469  Kristine Stolakis        None

## Issue 6
#### This involves data cleaning by working on the `year` column provided in the dataset to define and determine the category the movie falls into, either a completed or an ongoing movie

In [1148]:
#changing the column year to str type
df_movies['year']= df_movies['year'].astype(str)

In [1149]:
#extracting the years column 
df_movies['years'] = df_movies['year'].apply(lambda x: x.lstrip("(").rstrip(")"))

In [1150]:
#function to convert the years column to a list and specify the movie category
def dates_confirmn(x):
    dates = x.split("–")
    if len(dates) != 2:
        return 'completed_movie'
    elif dates[1] == " ":
        return 'ongoing_series'
    else:
        return 'completed_series'

In [1151]:
#applying the function
df_movies['movie_category'] = df_movies['years'].apply(dates_confirmn)

## Issue 7
#### Ratings : Filling the null values  in `ratings` with the average mean, this is because we have a low null values ratings compared to the filled ratings, i.e 1,820 null values compared to 8,179 filled values which is less than 23%

In [1152]:
#checking how many ratings rows are null
df_movies['rating'].isnull().sum()

1820

In [1153]:
df_movies['rating']

0       6.1
1       5.0
2       8.2
3       9.2
4       NaN
       ... 
9994    NaN
9995    NaN
9996    NaN
9997    NaN
9998    NaN
Name: rating, Length: 9999, dtype: float64

In [1154]:
#calculating the average mean of the ratings and fill the null values with the mean
df_movies['rating'] = df_movies['rating'].fillna(df_movies['rating'].mean()).round(1)


In [1155]:
#confirmining the null values ha
df_movies['rating'].isnull().sum()

0

## Issue 8
> Gross : The data cleaning process here is to fill the null values in `gross` column with zero as there are large null empty rows compared to the filled rows. 
- Then proceed to replace the $ dollar sign as this is not required in this data, before converting the rows with M symbol which represnts millions to  million values

In [1156]:
#confirming  the no of gross rows that are null
df_movies['gross'].isnull().sum()

9539

In [1157]:
#filling the null gross column with zero, this shows that no gross was allocated to the movie
df_movies['gross'] = df_movies['gross'].fillna(0)

In [1158]:
df_movies['gross'].isnull().sum()

0

In [1159]:
#convert to str datatype
df_movies['gross'] = df_movies['gross'].astype(str)

#extract and replace the $ values in the gross column
df_movies['gross']= df_movies['gross'].apply(lambda x: x.replace("$",""))

In [1160]:
#calling a million value convertion function
def to_millions(x):
    if 'M' in x:
        x = x.replace("M","")
        x = float(x)
        x = int(x * 1000000)
        return x
    else:
        return x
        

In [1161]:
#applying the funtion to convert the gross M to millions
df_movies['gross'] = df_movies['gross'].apply(to_millions)

In [1162]:
#converting the gross back to int and asserting the datatype
df_movies['gross'] = df_movies['gross'].astype(int)
print(df_movies['gross'].dtype)

int64


## Issue 9
#### Votes : Filling the `votes` null values with zero, take away the commas and then convert to the required datatype

In [1163]:
#filling the null votes with 0
df_movies['votes'] = df_movies['votes'].fillna(0)

In [1164]:
#convert to str datatype
df_movies['votes'] = df_movies['votes'].astype(str)

#replacing the commas with empty in the votes column
df_movies['votes'] = df_movies['votes'].apply(lambda x: x.replace(",", ""))

In [1165]:
#convert back to int 
df_movies['votes'] = df_movies['votes'].astype(int)

## Issue 10
#### Run_time : Converting the  `run_time` null values to zero

In [1166]:
#filling the null votes with 0
df_movies['run_time'] = df_movies['run_time'].fillna(0)

#convert to int type 
df_movies['run_time'] = df_movies['run_time'].astype(int)

In [1167]:
df_movies.dtypes

movies             object
year               object
genre              object
rating            float64
one_line           object
stars              object
votes               int64
run_time            int64
gross               int64
director           object
movies_star        object
years              object
movie_category     object
dtype: object

### Dropping Columns
> Unnecessary columns `year`, `stars` would be dropped off. New columns have been derived from these variables and hence, they are no longer required for further analysis

In [1168]:
#dropping the columns
df_movies.drop(['year','stars'], axis = 1, inplace = True)

In [1169]:
df_movies.head(50)

movies                       genre  \
0                                Blood Red Sky      Action,Horror,Thriller   
1          Masters of the Universe: Revelation  Animation,Action,Adventure   
2                             The Walking Dead       Drama,Horror,Thriller   
3                               Rick and Morty  Animation,Adventure,Comedy   
4                              Army of Thieves         Action,Crime,Horror   
5                                  Outer Banks          Action,Crime,Drama   
6              The Last Letter from Your Lover               Drama,Romance   
7                                       Dexter         Crime,Drama,Mystery   
8                            Never Have I Ever                      Comedy   
9                                 Virgin River               Drama,Romance   
10                         Gunpowder Milkshake   Action,Adventure,Thriller   
11                                     Lucifer         Crime,Drama,Fantasy   
12                           Fear Street: 1994        Drama,Horror,Mystery   
13                                    Sex/Life        Comedy,Drama,Romance   
14                       American Horror Story       Drama,Horror,Thriller   
15                              Grey's Anatomy               Drama,Romance   
16                                Breaking Bad        Crime,Drama,Thriller   
17                             The Good Doctor                       Drama   
18                                    Atypical                Comedy,Drama   
19                             Stranger Things        Drama,Fantasy,Horror   
20                           Fear Street: 1978        Drama,Horror,Mystery   
21                                     Friends              Comedy,Romance   
22                                   The Flash      Action,Adventure,Drama   
23                              Peaky Blinders                 Crime,Drama   
24      He-Man and the Masters of the Universe  Animation,Action,Adventure   
25                                   Chernobyl       Drama,History,Romance   
26                                     Vikings      Action,Adventure,Drama   
27                            Better Call Saul                 Crime,Drama   
28                                Supernatural        Drama,Fantasy,Horror   
29                               The Blacklist         Crime,Drama,Mystery   
30                           Fear Street: 1666              Horror,Mystery   
31                              Animal Kingdom                 Crime,Drama   
32                          Brooklyn Nine-Nine                Comedy,Crime   
33  NCIS: Naval Criminal Investigative Service          Action,Crime,Drama   
34                                     Kingdom        Action,Drama,History   
35                               Modern Family        Comedy,Drama,Romance   
36                                      Titans      Action,Adventure,Crime   
37                                 The Witcher    Action,Adventure,Fantasy   
38                               Downton Abbey               Drama,Romance   
39                          Shingeki no kyojin  Animation,Action,Adventure   
40                                       Suits                Comedy,Drama   
41                            La casa de papel        Action,Crime,Mystery   
42                              Schitt's Creek                      Comedy   
43                                   Outlander       Drama,Fantasy,Romance   
44                            The Last Kingdom        Action,Drama,History   
45                                Black Mirror       Drama,Sci-Fi,Thriller   
46                                       Ozark        Crime,Drama,Thriller   
47                                   The Crown     Biography,Drama,History   
48                                 Sweet Tooth      Action,Adventure,Drama   
49                            Army of the Dead         Action,Crime,Horror   

    rating                                           one_line

## Drop Duplicates
#### There are currently 431 rows movies that are duplicates, so I would be dropping them to know the exact number of movies we have.

In [1170]:
#finding out no of duplicates
df_movies.duplicated().value_counts()

False    9568
True      431
dtype: int64

In [1171]:
duplicated_df_movies = df_movies[df_movies.duplicated()]
duplicated_df_movies.head(5)

movies                      genre  rating    one_line  \
6833   Mighty Little Bheem  Animation,Short,Adventure     6.9  Add a Plot   
6835   Mighty Little Bheem  Animation,Short,Adventure     9.0  Add a Plot   
6836   Mighty Little Bheem  Animation,Short,Adventure     9.0  Add a Plot   
6837   Mighty Little Bheem  Animation,Short,Adventure     6.9  Add a Plot   
6838   Mighty Little Bheem  Animation,Short,Adventure     6.9  Add a Plot   

      votes  run_time  gross                                  director  \
6833      0         0      0  Rajiv Chilaka, Krishna Mohan Chintapatla   
6835      6         0      0  Rajiv Chilaka, Krishna Mohan Chintapatla   
6836      6         0      0  Rajiv Chilaka, Krishna Mohan Chintapatla   
6837      0         0      0  Rajiv Chilaka, Krishna Mohan Chintapatla   
6838      0         0      0  Rajiv Chilaka, Krishna Mohan Chintapatla   

     movies_star   years  movie_category  
6833        None  2019–   ongoing_series  
6835        None  2019–   ongoing_series  
6836        None  2019–   ongoing_series  
6837        None  2019–   ongoing_series  
6838        None  2019–   ongoing_series

In [1172]:
# Delete duplicate rows based on specific column 
df_movies = df_movies.drop_duplicates(subset=['movies'], keep=False)

In [1174]:
#the cleaned dataset
df_movies.head(5)

movies                       genre  rating  \
0                        Blood Red Sky      Action,Horror,Thriller     6.1   
1  Masters of the Universe: Revelation  Animation,Action,Adventure     5.0   
2                     The Walking Dead       Drama,Horror,Thriller     8.2   
3                       Rick and Morty  Animation,Adventure,Comedy     9.2   
4                      Army of Thieves         Action,Crime,Horror     6.9   

                                            one_line   votes  run_time  gross  \
0  A woman with a mysterious illness is forced in...   21062       121      0   
1  The war for Eternia begins again in what may b...   17870        25      0   
2  Sheriff Deputy Rick Grimes wakes up from a com...  885805        44      0   
3  An animated series that follows the exploits o...  414849        23      0   
4  A prequel, set before the events of Army of th...       0         0      0   

                director                                        movies_star  \
0        Peter Thorwarth  Peri Baumeister,Carl Anton Koch,Alexander Sche...   
1                   None  Chris Wood,Sarah Michelle Gellar,Lena Headey,M...   
2                   None  Andrew Lincoln,Norman Reedus,Melissa McBride,L...   
3                   None  Justin Roiland,Chris Parnell,Spencer Grammer,S...   
4  Matthias Schweighöfer  Matthias Schweighöfer,Nathalie Emmanuel,Ruby O...   

       years    movie_category  
0       2021   completed_movie  
1     2021–     ongoing_series  
2  2010–2022  completed_series  
3     2013–     ongoing_series  
4       2021   completed_movie